TicTacToe: Rspec tests

PhillipUg opened the pull request:
In this branch, we did the following:
- Added rspec tests for our game logic in the lib folder
- Slightly modified our classes to allow for testability

PhillipUg committed:
Initial commit. Initialized rspec

PhillipUg committed:
Added rspec formating option in .rspec file

PhillipUg committed:
setup pending tests for board.rb

PhillipUg committed:
refactored game object initialization

PhillipUg committed:
renamed 'player' argument into 'letter' to give it more contextual meaning

PhillipUg committed:
refactored ::valid_letter? method

adamclasic committed:
add relatives

PhillipUg committed:
refactored valid_position? method to take in Integer instead of String

PhillipUg committed:
reconfigured position variable to store Integer, instead of String

PhillipUg committed:
made Board class methods private

PhillipUg committed:
renamed 'no_winner?' method to 'draw?' for clarity

PhillipUg committed:
Board class methods changed back to public methods

PhillipUg committed:
Added test examples for Board and Game classes

PhillipUg committed:
resolved conflicts

PhillipUg committed:
Merge branch 'development' into rspec-tests

readme behind

PhillipUg committed:
updated readme

PhillipUg committed:
corrected rubocop errors

PhillipUg committed:
Merge branch 'rspec-tests' of github.com:adamclasic/tictactoe into rspec-tests

edited remote-readme

majovanilla commented on README.md:
```diff
@@ -144,8 +144,7 @@ To play this game you need to:
 ### Built With
 This project was built using these technologies.
 * Ruby
-* Rubocop
-* VsCode
+* Rspec
```
You also need to include a Testing section where you explain how to run the test suite.

majovanilla commented on spec/tictactoe_spec.rb:
```diff
@@ -0,0 +1,142 @@
+    context 'when num even' do
+      let(:num) { 0 }
+
+      it 'should check if number is even?' do
+        expect(num.even?).to eq(true)
+      end
```
Please test for the negative case, too.

majovanilla commented on spec/tictactoe_spec.rb:
```diff
@@ -0,0 +1,142 @@
+    context 'when num odd' do
+      let(:num) { 1 }
+
+      it 'should check if number is odd?' do
+        expect(num.odd?).to be true
+      end
```
Check for the negative case, too.

majovanilla commented on spec/tictactoe_spec.rb:
```diff
@@ -0,0 +1,142 @@
+  describe '::valid_letter?' do
+    subject { Game.valid_letter?(letter) }
+    let(:letter) { 'D' }
+    it 'should accept a char (string) as an argument' do
+      expect(letter).to be_kind_of String
```
You should also check the negative case (not be) if the user inputs something else than a string.

majovanilla commented on spec/tictactoe_spec.rb:
```diff
@@ -0,0 +1,142 @@
+      it { should be true }
+    end
+
+    context 'when not a valid position' do
+      it { should be false }
```
You are not testing anything here. You should replicate the player's behavior (use the subject variable) and have a taken position and then try to select it.
If you change the `false` to `true` it still works.

majovanilla commented on spec/tictactoe_spec.rb:
```diff
@@ -0,0 +1,142 @@
+    end
+
+    context 'when a valid position' do
+      let(:position) { 3 }
+      it { should be true }
```
You are not testing the method here. You should `expect(subject)` to be or do something, but you are not. Please fix this.

Here you can find more info on [subject](https://relishapp.com/rspec/rspec-core/v/3-8/docs/subject/explicit-subject).

majovanilla commented on spec/tictactoe_spec.rb:
```diff
@@ -0,0 +1,142 @@
+      expect(game.table[position]).to eq(current_player_letter)
+    end
+
+    it 'should return an array' do
+      should be_kind_of Array
+    end
```
To make your test better, you have to check if the array is the one you are expecting, after some player's moves, see if it includes them.

majovanilla commented on spec/tictactoe_spec.rb:
```diff
@@ -0,0 +1,142 @@
+  end
+
+  describe '#add_to_table' do
+    subject { game.add_to_table(position) }
```
You are not using `subject` in this test. You should delete it or use it in your methods. Check the resource linked above if you have doubts about how `subject` works.

majovanilla commented on spec/tictactoe_spec.rb:
```diff
@@ -0,0 +1,142 @@
+
+    let(:current_player_letter) { 1 }
+    it 'should set table position to current player letter' do
+      game.table[position] = current_player_letter
+      expect(game.table[position]).to eq(current_player_letter)
```
You are not testing your method here. You are assigning the value of  ` game.table[position] ` directly, so it will always be equal to what you expect! You should run the actual method and then compare for equality.

majovanilla commented on spec/tictactoe_spec.rb:
```diff
@@ -0,0 +1,142 @@
+    subject { game.add_to_table(position) }
+    let(:position) { 2 }
+
+    it 'should accept an Integer (position) as an argument' do
+      expect(position).to be_kind_of Integer
+    end
```
Check for the negative case.

majovanilla reviewed (changes requested):
Hi team!!

## You did a *great* job on this project! 💜 

There are still some changes 👇 you need to implement to approve this request 🔍

- The best way to organize your test suite is to follow the same structure you have on your lib folder: create different testing files for each class and another one for the methods you are testing on the `main.rb` Maybe right now you don't have a thousand lines of code, but it's always better to implement the best practices for when you build larger projects.

## Please, make all REQUIRED changes on the commented code below and submit your request again! 😀

If you have any questions, feel free to send me a message through Slack (@majo)
I would be happy to help! 👩‍💻 


Happy coding 🤓

--Maria Reyes

PhillipUg committed:
updated readme

Added testing section

PhillipUg commented:
Thanks @majovanilla for taking the time to review our work. We will make the necessary changes and make another PR

PhillipUg committed:
made specs more modular. Wrote negative cases

PhillipUg committed:
corrected rubocop error

PhillipUg commented:
## We have made the following improvements:
  - added testing section in README
  - added negative cases as suggested in the previous review
  - made our specs more modular by separating specs for each class
  - Corrected all the errors flagged by the previous reviewer
  
However, we have discussed with the previous reviewer and agreed that some of the cases flagged in the previous review might have been flagged mistakenly. Like the usage of of `subject`.

whiz25 reviewed (changes requested):
## Good Job :+1:
You have made the suggested changes by the previous reviewer but on testing, there is a failing edge case. 
`NoMethodError: undefined method current_player_letter=`
On line 92 of `spec/game_spec.rb` file.
Please fix that and send for another review

### Martin Afani

PhillipUg committed:
made current player letter variable accessible

PhillipUg commented:
Thanks @whiz25 Lets us correct the error

PhillipUg commented:
## Corrected:
  - failing test case cited by previous reviewer

kobiyoyo reviewed (changes requested):
# Project Approved :thumbsup:
## Hello Team,  You have done a great job with this project. :clap: 
#### You can go ahead and merge !!!

kobiyoyo reviewed (approved):
Approved

PhillipUg commented:
Thanks @kobiyoyo
